runtime/src/main/java/com/dylibso/chicory/runtime/MStack.java

```diff
@@ -1,42 +1,797 @@
+     */
+    public boolean checkTypes(List<ValueType> types) {
+        int cnt = types.size();
+        int start = depth - cnt;
+        if (start < base) {
+            throw underflow();
+        }
+        for (int i = 0; i < cnt; i++) {
+            if (!idToType(this.types[start + i]).equals(types.get(i))) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Unwind the stack to the given depth.
+     *
+     * @param depth the depth to unwind to
+     */
+    public void resetDepth(int depth) {
+        if (base > depth || depth > this.depth) {
+            throw underflow();
+        }
+        drop(this.depth - depth);
+    }
+
+    /**
+     * Drop some values from the top of the stack.
+     *
+     * @param count the number of values to drop
+     */
+    public void drop(int count) {
+        drop(count, 0);
+    }
+
+    /**
+     * Drop some values from the stack.
+     *
+     * @param dropCount the number of values to drop
+     * @param skip the number of values to skip before dropping values
+     */
+    public void drop(int dropCount, int skip) {
+        if (dropCount < 0 || skip < 0) {
+            throw new IllegalArgumentException("dropCount and skip must not be less than 0");
+        }
+        if (dropCount == 0) {
+            // do nothing
+            return;
+        }
+        if (depth - dropCount - skip < base) {
+            throw underflow();
+        }
+        // compute the number of entries to skip on each substack
+        int skipWords = 0;
+        int skipRefs = 0;
+        for (int i = 0; i < skip; i++) {
+            ValueType tos = idToType(types[depth - i - 1]);
+            skipWords += wordCount(tos);
+            skipRefs += refCount(tos);
+        }
+        // compute the number of entries to drop on each substack
+        int dropWords = 0;
+        int dropRefs = 0;
+        for (int i = 0; i < dropCount; i++) {
+            ValueType tos = idToType(types[depth - i - skip - 1]);
+            dropWords += wordCount(tos);
+            dropRefs += refCount(tos);
+        }
+        if (skip > 0) {
+            // preserve the top N value types
+            System.arraycopy(types, depth - skip, types, depth - dropCount - skip, skip);
+            // words
+            System.arraycopy(
+                    words,
+                    wordDepth - skipWords,
+                    words,
+                    wordDepth - dropWords - skipWords,
+                    skipWords);
+            // refs
+            System.arraycopy(
+                    refs, refDepth - skipRefs, refs, refDepth - dropRefs - skipRefs, skipRefs);
+        }
+        // clear refs for GC
+        Arrays.fill(refs, refDepth - dropRefs, refDepth, null);
+        // finally, shrink the stacks
+        depth -= dropCount;
+        wordDepth -= dropWords;
+        refDepth -= dropRefs;
     }
```

Comment:
This is going to take some time for me to wrap my head around what is happening in this class. I understand that this level of control should make some performance improvements, but we're sacrificing the simplicity and safety of just using a standard type like Stack or ArrayDeque. It's hard to do a cost benefit-analysis on this kind of change, but it feels a bit premature before we have any users or the AOT compilation started.

Reply:
The gist if it is to copy the "keep" values to the lower array position, which is calculated arithmetically above by counting the number of substack slots. Since we have separate array substacks for the type ID, integral words, and ref types, we have to do three copies to copy down each "keep" region, and we also have to fill in the abandoned spot in the refs substack to avoid holding strong references to ref values after they are discarded.

Reply:
> It's hard to do a cost benefit-analysis on this kind of change, but it feels a bit premature before we have any users or the AOT compilation started.

I'm of the opinion that this is better done sooner rather than later - this puts a lot less pressure on the GC, and probably also helps a fair amount with CPU cache locality and memory bandwidth overhead. It also has fairly high risk of impacting other foundational classes (such as `Machine`, as happened in this PR). I've generally found that in interpreters and compilers, foundational changes are easier to make sooner rather than later.

I agree that there are plenty of optimizations that are best left for later, but I'm in favor of picking off low-hanging fruit (especially if it has large and generally beneficial impact on the codebase) sooner rather than later. My concern with only writing code that pays little to no attention to performance is that if users start picking it up now it might get a reputation of being "slow," and that can be hard to shake.

For AOT, I think this is either irrelevant or minimal impact - it may be useful to use `MStack` for some type tracking tasks during bytecode generation, but I'm not expecting impact beyond that.

Reply:
> a bit premature before we have any users

I'm aligned, we agreed on completing full support for the WASM specification before we dig into invasive performance improvements.
Currently, we are barely starting to crack validation.

> I'm in favor of picking off low-hanging fruit 

LOC is not always a good indication of the complexity of something, but, at least, we need to review and go through 1700+ lines to incorporate this change:
![Screenshot 2024-02-28 at 09 40 44](https://github.com/dylibso/chicory/assets/5792097/b7db67ae-32bd-4049-9333-8be506a34219)
calling it a "low-hanging fruit" sounds a bit of a stretch to me.

Reply:
With respect, I don't think it is likely that the full WASM spec is supportable under the current architecture without incurring substantial technical debt. Speaking as someone who has worked on *hundreds* of OSS projects, it is absolutely and unequivocally *always* best to design for the future. For example, I can say, without reservation, that in the future having box objects for every value is going to be a significant problem that will be much harder to solve later than it is to solve now.

This patch is as much about having a clean future-ready API which supports all of the spec types as it is about performance.

This patch is also about simplicity. Yes, you are correct that the change adds a lot of lines, because the old API usages are replaced with the new API. `cloc` tells me that the `MStack` implementation itself is 613 lines of code (of which 10% is dedicated to just having a nice `toString` method) plus 345 lines of comments. The tests add 369 lines (tests are good, right?). The remaining LOC count is reworking and simplifying the existing code to use the new implementation efficiently. It's disingenuous to characterize the patch as complex based on +1722 additions.

What you get for 613 lines is a clean and efficient stack implementation which supports every existent WASM type with minimal memory overhead. Simplicity isn't about avoiding complexity. It's about *managing* complexity by encapsulating it behind simple interfaces, and that's exactly what this patch strives to do. Try not to think about how complex the implementation itself is, but how simple the API it presents is. With this API you can push and pop each primitive type, and even vector types, without additional box objects in a way that is idiomatic and intuitive. It also paves the way for call stack frames, which is something that will surely be much harder to implement as time goes on.

A system with clean APIs and strong encapsulation around arbitrary complexity is, by any reasonable calculation, simpler overall than a system which avoids any kind of complexity but also lacks encapsulation. It's never too early to define strong encapsulation, and the cost of not doing so is often very high.